proposals/Classes.md

```diff
@@ -0,0 +1,309 @@
+
+#### Defining functions
+
+```
+function<void> ReturnsVoid() {
+  this.someInt = 2
+}
+
+function<int> ReturnsInt(int a) {
+  return this.someInt + a + 1
+}
+
+static function<bool> DoSomething() {
+  return true
+}
+```
+
+### `property`: A property
+
+A property defines a named field on the class which executes logic on `get` (and optionally, `set`).
+If a property is not marked `static`, the `this` implicit value inside the scope of property provides access to the instance of the class.
+
+Properties must have a `get`, but `set` is optional (if the property is treated as a read-only or an inferred value).
+
+Inside a `set` block, an implicit `value` value is added with the value specified in the set expression.
+
+Properties **cannot** be of type `void`.
+
+#### Defining properties
+
+```
+property<int> SomeInt {
+  get { return this.someInt }
+}
+
+property<int> SomeOtherInt {
+  get { return this.someInt2 }
+  set { this.someInt2 = value }
+}
+```
+
+
+### `operator`: An operator
+
+An operator defines how an operator (+, -, ==, etc) works over a class of that type.
+
+Operators are always treated as **static** (even though they don't have the `static` keyword).
+
+Binary Operators must always take in two parameters, each a nullable value of the parent class,
+representing the left and right hand side of the expression.
+
+Unary Operators must always take in one parameter, a nullable value of the parent class. 
+
+#### Defining operators
+
+```
+operator plus (SomeClass? first, SomeClass? second) {
+  first = first ?? SomeClass.WithValue(0)
+  second = second ?? SomeClass.WithValue(0)
+  return SomeClass.WithValue(first.someInt + second.someInt)
+}
+
+operator not (SomeClass? value) {
+  return SomeClass.WithValue((value?.someInt * -1) ?? 0)
+}
+```
+
+
+### Combining classes (or: why no inheritance?)
+
+[Inheritance](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)) as found in
+traditional OOP languages presents a number of major issues around code reuse and principles of least
+surprise. Multiple inheritance can cause major headaches about reasoning ("diamond pattern") while 
+single inheritance can be extremely limiting when attempting to represent a type that operates as
+two different "kinds" of objects.
+
+#### Structural Subtyping
+
+For this project, we propose instead a form of **structural** subtyping (somewhat like Go) to
+allow for code use and implict [**interfaces**](Interfaces.md) to provide for sharing and substitution
+of like-kind types.
+
+Structural subtyping allows one class to share the code of one (or more) "parent" classes:
+
+```
+class AddClass {
+   function<int> Add(int first, int second) {
+     return first + second
+   }
+}
+
+class SubtractClass {
+   function<int> Subtract(int first, int second) {
+     return first - second
+   }
+}
+
+class AddSubtractClass : AddClass + SubtractClass {
```

Comment:
Two options:
`AddSubtractClass : AddClass + SubtractClass`
`AddSubtractClass : AddClass, SubtractClass`

I tend to prefer the `+`, as it clearly indicates that this is a combination, rather than an override of one by the other (although, that's how shadowing works as seen below)